src/analysis.zig

```diff
@@ -401,8 +361,89 @@ pub fn nodeToString(tree: *ast.Tree, node: *ast.Node) ?[]const u8 {
         },
         else => {
             std.debug.warn("INVALID: {}\n", .{node.id});
-        }
+        },
     }
-    
+
     return null;
 }
+
+pub const CompletionContext = struct {
+    pub const Id = enum {
+        builtin,
+        var_access,
+        field_access,
+        enum_literal,
+        container_field,
+        empty,
+        none,
+    };
+
+    id: Id,
+    node: ?*ast.Node = null,
+
+    fn toId(self: CompletionContext, id: Id) CompletionContext {
+        return .{ .id = id, .node = self.node };
+    }
+};
+
+fn nodeContainsSourceIndex(tree: *ast.Tree, node: *ast.Node, source_index: usize) bool {
+    const first_token = tree.tokens.at(node.firstToken());
+    const last_token = tree.tokens.at(node.lastToken());
+    return source_index >= first_token.start and source_index <= last_token.end;
+}
+
+fn visitNodesAndFindCompletionContext(tree: *ast.Tree, node: *ast.Node, source_index: usize) CompletionContext {
+    switch (node.id) {
+        .Identifier => {
+            var cc = CompletionContext{ .id = .var_access, .node = node };
+
+            if (nodeToString(tree, node).?[0] == '@') {
+                cc.id = .builtin;
+            }
+            return cc;
+        },
+        .InfixOp => {
+            const infix_op = node.cast(ast.Node.InfixOp).?;
+            if (infix_op.op == .Period and nodeContainsSourceIndex(tree, infix_op.rhs, source_index)) {
+                return .{ .id = .field_access, .node = node };
+            }
+        },
+        .BuiltinCall => return .{ .id = .builtin, .node = node },
+        .EnumLiteral => return .{ .id = .enum_literal, .node = node },
+        else => {},
+    }
+
+    var node_idx: usize = 0;
+    while (node.iterate(node_idx)) |child| : (node_idx += 1) {
+        if (!nodeContainsSourceIndex(tree, child, source_index)) continue;
```

Comment:
This won't stop us from iterating over all the children of this node.
If we don't want to iterate over all the child nodes and collect analytics at the same time (which I think we should do), we need to make our own iterator which will skip all the child nodes if the node doesn't contain our source index.

Reply:
Collecting the analytics at the same time as we are iterating should also make collecting all in scope variables quite easy.

Reply:
Yes, I agree that if we are going to be visiting the nodes it seems like a waste not to collect symbol tables and other information.  

Can you elaborate on the custom iterator?  
This loop will iterate over the children until it finds the child that contains the cursor, it will not iterate over the rest of the children.

Reply:
```zig
test "a" {
    // We will iterate all the nodes in here even though there
    // is no way any of them contain our index since the test 
    // declaration didn't contain it.
}
test "b" {
    std.<cursor here>
}
```

Reply:
How is this the case?  
We enter `visitNodesAndFindCompletionContext` with `node == &tree.root_node.base`, `iterate(0)` will yield the first `TestDecl`, which has `firstToken() == test` and `lastToken() == body_node.lastToken() == }`, so `nodeContainsSourceIndex` will return false and the loop will continue and yield the second `TestDecl`.  
We only recursively visit when `nodeContainsSourceIndex` is true.

Reply:
Ah, my bad. I was thinking of how the `Node.dump` function worked but I just noticed it calls `dump` on all the child nodes instead of `iterate` doing so.

Reply:
Cool, I was wondering if I missed some `iterate` that does a deep iteration :P